packages/liquidator/src/liquidationStrategy.js

```diff
@@ -0,0 +1,223 @@
+      syntheticTokenBalance,
+      positionTokens: position.numTokens,
+      empMinSponsorSize,
+      maxTokensToLiquidateWei
+    });
+
+    // check if we should try to liquidate this position
+    if (shouldLiquidate({ tokensToLiquidate })) {
+      // We should liquidate the entire position using our capital - wdf
+      return createLiquidationParams({
+        sponsor: position.sponsor,
+        tokensToLiquidate,
+        currentBlockTime,
+        maxCollateralPerToken
+      });
+    }
+    // we dont liquidate, dont return anything. Probably because amount to liquidate is 0
+  }
+
+  // Pure function which just calculates how far a timer has progressed between 0 and 100
+  // duration - total duration of the liveness in seconds
+  // end - end time that withdraw passes in unix seconds ts
+  // now - current block time in unix seconds ts
+  function withdrawProgressPercent(duration, end, now) {
+    const elapsed = end - now;
+    if (elapsed <= 0) return 100;
+    if (elapsed >= duration) return 0;
+    return 100 * (1 - elapsed / duration);
+  }
+
+  // Should we try to delay this withdrawal. Update this logic to change conditions to run delay strategy.
+  function shouldLiquidateMinimum({ position, syntheticTokenBalance, empMinSponsorSize, currentBlockTime }) {
+    assert(defenseActivationPercent >= 0, "requires defenseActivationPercent");
+    syntheticTokenBalance = toBN(syntheticTokenBalance);
+    empMinSponsorSize = toBN(empMinSponsorSize);
+
+    // this position does not have a withdraw pending
+    if (position.withdrawalRequestPassTimestamp == 0) return false;
+    // withdraw has passed liveness
+    if (position.withdrawalRequestPassTimestamp <= currentBlockTime) return false;
+    // our synthetic balance is less than the amount required to extend deadline
+    if (syntheticTokenBalance.lt(empMinSponsorSize)) return false;
+    // we have enough to fully liquidate position, so do not do the minimum
+    if (syntheticTokenBalance.gte(toBN(position.numTokens))) return false;
+    // position cant go below the minimum emp sponsor size
+    if (
+      toBN(position.numTokens)
+        .sub(empMinSponsorSize)
+        .lt(empMinSponsorSize)
+    )
+      return false;
+    // liveness timer on withdraw has not passed time
+    if (
+      withdrawProgressPercent(withdrawalLiveness, position.withdrawalRequestPassTimestamp, currentBlockTime) <
+      parseFloat(defenseActivationPercent)
+    )
+      return false;
+
+    // all conditions passed and we should minimally liquidate to extend timer
+    return true;
+  }
+
+  // Any new constraints can be added here, but for now
+  // only dont liquidate if the amount to liquidate is <=0
+  function shouldLiquidate({ tokensToLiquidate }) {
+    tokensToLiquidate = toBN(tokensToLiquidate);
+    // nothing to liquidate
+    if (tokensToLiquidate.lte(toBN(0))) return false;
+    // any other conditions we do try to liquidate
+    return true;
+  }
+
+  return {
```

Comment:
I like the idea of a liquidator "strategy". Have you considered building a standard interface or expected functions for a liquidator strategy so these could be plugged in and swapped out in a general way?

Reply:
Agreed that the concept of a strategy is awesome

Reply:
I think its possible to do this. The ideal interface is to pass in a position and have it spit out a liquidation size, or liquidation arguments. The difficulty is that theres a bunch of tangential state the strategy could or should take into consideration which would need to be standardized across all the strategies. That is more of the challenge, ie, what state can be introduced at strategy creation, vs what state needs to be passed in with each position, and can this all be the same per strategy.

Theres also the issue I ran into with running multiple strategies, I dont think running multiple strategies simultaneously is really possible without a new pattern or abstraction which manages how the strategies interact. Currently I think you would need to only run 1 at a time until there's some higher level logic choosing which strategy to employ on a position.

Reply:
No need to do this now, but I could imagine a small riff on what you're saying where you basically reduce across all strategies:
```
const liquidationArgs = strategies.reduce((currentArgs, strategy) => {
  return strategy.processPosition(currentArgs);
});
```

There is definitely ordering dependence, but you would basically want to work from least restrictive to most restrictive. However, I think it would be helpful to have another strategy or two before we understand what the interaction structure should be.

Reply:
yea the pattern you are proposing is quite similar to how express processes http requests. You define handlers that get called in order, and as the requests fall through you can mutate or add additional data in each handler. I could see it working that way under some strict assumptions about the strategies, but it does create a situation where the strategies can become dependent on each other, or have to have knowledge of each other, which would be nice to avoid somehow.

Reply:
Yeah, totally agree^